src/sio.ts

```diff
@@ -64,129 +70,152 @@ export class RPSIO {
-        this.divDividend = value;
-        this.updateHardwareDivider(true);
-        break;
-      case DIV_UDIVISOR:
-        this.divDivisor = value;
-        this.updateHardwareDivider(false);
-        break;
-      case DIV_SDIVISOR:
-        this.divDivisor = value;
-        this.updateHardwareDivider(true);
-        break;
-      case DIV_QUOTIENT:
-        this.divQuotient = value;
-        this.divCSR = 0b11;
-        break;
-      case DIV_REMAINDER:
-        this.divRemainder = value;
-        this.divCSR = 0b11;
-        break;
+    if (offset < SPINLOCK0) {
+      //not a spinlock
+      switch (offset) {
+        case GPIO_OUT:
+          this.gpioValue = value & GPIO_MASK;
+          break;
+        case GPIO_OUT_SET:
+          this.gpioValue |= value & GPIO_MASK;
+          break;
+        case GPIO_OUT_CLR:
+          this.gpioValue &= ~value;
+          break;
+        case GPIO_OUT_XOR:
+          this.gpioValue ^= value & GPIO_MASK;
+          break;
+        case GPIO_OE:
+          this.gpioOutputEnable = value & GPIO_MASK;
+          break;
+        case GPIO_OE_SET:
+          this.gpioOutputEnable |= value & GPIO_MASK;
+          break;
+        case GPIO_OE_CLR:
+          this.gpioOutputEnable &= ~value;
+          break;
+        case GPIO_OE_XOR:
+          this.gpioOutputEnable ^= value & GPIO_MASK;
+          break;
+        case GPIO_HI_OUT:
+          this.qspiGpioValue = value & GPIO_MASK;
+          break;
+        case GPIO_HI_OUT_SET:
+          this.qspiGpioValue |= value & GPIO_MASK;
+          break;
+        case GPIO_HI_OUT_CLR:
+          this.qspiGpioValue &= ~value;
+          break;
+        case GPIO_HI_OUT_XOR:
+          this.qspiGpioValue ^= value & GPIO_MASK;
+          break;
+        case GPIO_HI_OE:
+          this.qspiGpioOutputEnable = value & GPIO_MASK;
+          break;
+        case GPIO_HI_OE_SET:
+          this.qspiGpioOutputEnable |= value & GPIO_MASK;
+          break;
+        case GPIO_HI_OE_CLR:
+          this.qspiGpioOutputEnable &= ~value;
+          break;
+        case GPIO_HI_OE_XOR:
+          this.qspiGpioOutputEnable ^= value & GPIO_MASK;
+          break;
+        case DIV_UDIVIDEND:
+          this.divDividend = value;
+          this.updateHardwareDivider(false);
+          break;
+        case DIV_SDIVIDEND:
+          this.divDividend = value;
+          this.updateHardwareDivider(true);
+          break;
+        case DIV_UDIVISOR:
+          this.divDivisor = value;
+          this.updateHardwareDivider(false);
+          break;
+        case DIV_SDIVISOR:
+          this.divDivisor = value;
+          this.updateHardwareDivider(true);
+          break;
+        case DIV_QUOTIENT:
+          this.divQuotient = value;
+          this.divCSR = 0b11;
+          break;
+        case DIV_REMAINDER:
+          this.divRemainder = value;
+          this.divCSR = 0b11;
+          break;
+      }
+    }
+    if (offset >= SPINLOCK0 && offset <= SPINLOCK31) {
```

Comment:
same comment as above - moving the `if` above the `switch` and adding a `return` so it never gets to the switch in case of a spinlock.